### ISSUE TITLE
Visual mode not correctly set up on a keymap set up through lua api

### ISSUE THREAD

**Issue by hkupty** [neovim/neovim#17735]
### Neovim version (nvim -v)

v0.7.0-dev+1258-gbef7552f7

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

Linux 5.16.13-arch1-1

### Terminal name/version

kitty 0.24.4

### $TERM environment variable

xterm-256color

### Installation

AUR

### How to reproduce the issue

On a lua config:

```lua
vim.keymap.set({'v'}, '<space><space>', function()
  print(vim.inspect{
      vim.fn.visualmode(),
      vim.fn.getpos("'<'"),
      vim.fn.getpos("'>'"),
      vim.fn.line('v'),
      vim.fn.col('v')
})
end, { silent = true })
```

Then, on visual selection, run `<space><space>`

### Expected behavior

Inspect result would show correct visual mode as well as correct positions for `getpos` and `col/line` pair.

### Actual behavior

Visual mode is kept, not being able to extract any meaningful values from the visual mode. On the second call to that mapping we can get the first set of values.

Even if force breaking out of visual mode by adding a `vim.api.nvim_input("<ESC>")` line before the inspect function call, it will still show the wrong values, leading me to assume the visual values are set before the keymap is executed.

**Comment by zeertzjq** [comment]
Using `vim.api.nvim_input("<ESC>")` definitely won't work here. You need to use `vim.api.nvim_feedkeys(vim.api.nvim_replace_termcodes('<Esc>', false, true, true), 'nx', false)`. You may also need to set `vim.o.showmode = false`.

**Comment by hkupty** [comment]
Right. I guess the issue is that I can't get selection information from within the mapping while in visual mode unless I break out of visual. Not really sure what is best, if keeping the visual selection or not.

> Using `vim.api.nvim_input("<ESC>")` definitely won't work here.
I thought it would replace the termcode for me. 😓
I guess an API function for breaking out of visual would be preferred, but using feedkeys should work then.

**Comment by zeertzjq** [comment]
`nvim_input()` will replace the termcode, but it will just enqueue the key, not execute it.

**Comment by justinmk** [comment]
> You need to use `vim.api.nvim_feedkeys(vim.api.nvim_replace_termcodes('<Esc>', false, true, true), 'nx', false)`. You may also need to set `vim.o.showmode = false`.

Adding that before the inspect() call, works. `feedkeys(..., 'nx')` is the general practice to execute input immediately.

# Proposal

It might be useful if `interrupt()` accepted a flag that made it useful in more situations, like this.

**Comment by justinmk** [comment]
> the issue is that I can't get selection information from within the mapping while in visual mode

`getregionpos()` can be used for this now.
